File: examples/E08-templates/input/src/generico.cpp
#include "generico.hpp"

template<>
std::string dobro<std::string>(std::string valor) {
    return valor + " (dobrado)";
}

template<>
std::string Caixa::pega<std::string>(const std::string &chave) const {
    return "valor:" + chave;
}

template<>
int Caixa::pega<int>(const std::string &chave) const {
    return static_cast<int>(chave.length());
}

template<>
bool Caixa::tem<std::string>(const std::string &chave) const {
    return chave.length() > 0;
}

int testarDobroInt() {
    return dobro(5);
}

double testarDobroDouble() {
    return dobro(2.5);
}

std::string testarDobroString() {
    return dobro(std::string("oi"));
}

std::string testarCaixaString() {
    Caixa c;
    return c.pega<std::string>("teste");
}

int testarCaixaInt() {
    Caixa c;
    return c.pega<int>("cinco");
}

bool testarCaixaTem() {
    Caixa c;
    return c.tem<std::string>("algo");
}

File: examples/E08-templates/expected/lib/generico.dart
import 'dart:convert';

class Caixa {
  Caixa();

  Caixa.syntaxBridgeCopyOf(Caixa other) {}

  String pegaString(String chave) {
    return 'valor:' + chave;
  }

  int pegaInt(String chave) {
    return utf8.encode(chave).length;
  }

  bool temString(String chave) {
    return utf8.encode(chave).length > 0;
  }
}

String dobroString(String valor) {
  return valor + ' (dobrado)';
}

double dobroDouble(double valor) {
  return valor + valor;
}

int dobroInt(int valor) {
  return valor + valor;
}

int testarDobroInt() {
  return dobroInt(5);
}

double testarDobroDouble() {
  return dobroDouble(2.5);
}

String testarDobroString() {
  return dobroString('oi');
}

String testarCaixaString() {
  Caixa c = Caixa();
  return c.pegaString('teste');
}

int testarCaixaInt() {
  Caixa c = Caixa();
  return c.pegaInt('cinco');
}

bool testarCaixaTem() {
  Caixa c = Caixa();
  return c.temString('algo');
}
